app/src/rgb_underglow.c

```diff
@@ -0,0 +1,307 @@
+
+        pixels[i] = hsb_to_rgb(hsb);
+    }
+
+    state.animation_step += state.animation_speed * 10;
+    
+    if (state.animation_step > 2400) {
+        state.animation_step = 0;
+    }
+}
+
+static void zmk_rgb_underglow_effect_spectrum()
+{
+    for (int i=0; i<STRIP_NUM_PIXELS; i++)
+    {
+        int hue = state.animation_step;
+        int sat = state.saturation;
+        int brt = state.brightness;
+
+        struct led_hsb hsb = { hue, sat, brt };
+
+        pixels[i] = hsb_to_rgb(hsb);
+    }
+
+    state.animation_step += state.animation_speed;
+    state.animation_step = state.animation_step % 360;
+}
+
+static void zmk_rgb_underglow_effect_swirl()
+{
+    for (int i=0; i<STRIP_NUM_PIXELS; i++)
+    {
+        int hue = (360 / STRIP_NUM_PIXELS * i + state.animation_step) % 360;
+        int sat = state.saturation;
+        int brt = state.brightness;
+
+        struct led_hsb hsb = { hue, sat, brt };
+
+        pixels[i] = hsb_to_rgb(hsb);
+    }
+
+    state.animation_step += state.animation_speed * 2;
+    state.animation_step = state.animation_step % 360;
+}
+
+static void zmk_rgb_underglow_tick(struct k_work *work)
+{
+    switch (state.current_effect)
+    {
+        case UNDERGLOW_EFFECT_SOLID:
+            zmk_rgb_underglow_effect_solid();
+            break;
+        case UNDERGLOW_EFFECT_BREATHE:
+            zmk_rgb_underglow_effect_breathe();
+            break;
+        case UNDERGLOW_EFFECT_SPECTRUM:
+            zmk_rgb_underglow_effect_spectrum();
+            break;
+        case UNDERGLOW_EFFECT_SWIRL:
+            zmk_rgb_underglow_effect_swirl();
+            break;
+    }
+
+    led_strip_update_rgb(led_strip, pixels, STRIP_NUM_PIXELS);
+}
+
+K_WORK_DEFINE(underglow_work, zmk_rgb_underglow_tick);
+
+static void zmk_rgb_underglow_tick_handler(struct k_timer *timer)
+{
+    k_work_submit(&underglow_work);
+}
+
+K_TIMER_DEFINE(underglow_tick, zmk_rgb_underglow_tick_handler, NULL);
+
+static int zmk_rgb_underglow_init(struct device *_arg)
+{
+	led_strip = device_get_binding(STRIP_LABEL);
+	if (led_strip) {
+		LOG_INF("Found LED strip device %s", STRIP_LABEL);
+	} else {
+		LOG_ERR("LED strip device %s not found", STRIP_LABEL);
+		return -EINVAL;
+	}
+
+    state = (struct rgb_underglow_state){
+        hue: 0,
+        saturation: 100,
+        brightness: 100,
+        animation_speed: 3,
+        current_effect: 0,
+        animation_step: 0,
+        on: true
+    };
+
+    k_timer_start(&underglow_tick, K_NO_WAIT, K_MSEC(50));
+
+    return 0;
+}
+
+int zmk_rgb_underglow_cycle_effect(int direction)
+{
+    if (state.current_effect == 0 && direction < 0) {
+        state.current_effect = UNDERGLOW_EFFECT_NUMBER - 1;
+        return 0;
+    }
+
+    state.current_effect += direction;
+
+    if (state.current_effect >= UNDERGLOW_EFFECT_NUMBER) {
+        state.current_effect = 0;
+    }
+    
+    state.animation_step = 0;
+
+    return 0;
+}
+
+int zmk_rgb_underglow_toggle()
+{
+    state.on = !state.on;
```

Comment:
Something I missed in the first pass. There needs to be some early checks in any of these functions to be sure we actually have an `led_strip` device.

If init failed... the behaviors still can be triggered by the keymap, or other folks might try to call into this, so checks in various "entrypoints" to this should assert the `led_strip` is there and exit early otherwise.

Reply:
Should the behaviors file still be gated by the ZMK_RGB_UNDERGLOW Kconfig value in this case? Also technically the only behavior function that interacts directly with the led strip would be zmk_rgb_underglow_toggle. Should this be the only one that needs to make this check, or should I check in all of the non-static functions so I can make a special return code?

Reply:
Probably good for all to check. Even though they don't technically interact with the device, the behaviour still failed to perform the requested operation. That should be reported so it can bubble up to higher layers. (E.g. later we might do some reporting to OLEDs, or elsewhere, etc.

Reply:
And it should still be gated, so that folks not using underglow don't need that code increasing the size of their final image.